README.md

```diff
@@ -547,6 +541,92 @@ In this case:
+
+An additional API that would be helpful, both for cases like these and more generally, would be one for [setting and getting the sequential focus navigation start point](https://github.com/whatwg/html/issues/5326). Especially for the custom traversals case, this would give even higher-fidelity focus restoration. (But that proposal is separate from app history.)
+
+We can also extend the `focusReset` option with other behaviors in the future. Here are a couple which have been proposed, but we are not planning to include in the initial version unless we get strong developer feedback that they would be helpful:
+
+- `e.transitionWhile(promise, { focusReset: "immediate" })`: immediately resets the focus to the `<body>` element, without waiting for the promise to settle.
+- `e.transitionWhile(promise, { focusReset: "two-stage" })`: immediately resets the focus to the `<body>` element, and then has the same behavior as `"after-transition"`.
+
+#### Scroll position restoration
+
+A common paint point for web developers is scroll restoration during traversal (back/forward) navigations. The essential problem is that scroll restoration happens unpredictably, and often at the wrong times. For example:
+
+- The browser tries to restore the user's scroll position, but the application logic is still setting up the DOM and the relevant elements aren't ready yet.
+- The browser tries to restore the user's scroll position, but the page's contents have changed and scroll restoration doesn't work that well. (For example, going back to a listing of files in a shared folder, after a different user deleted a bunch of the files.)
+- The application needs to perform some measurements in order to do a proper transition, but the browser does scroll restoration during the transition, which messes up those measurements. ([Demo of this problem](https://nifty-blossom-meadow.glitch.me/legacy-history/transition.html): notice how when going back to the grid view, the transition sends the square to the wrong location.)
+
+Currently the browser provides two options: performing scroll restoration automatically, or disabling it entirely with `history.scrollRestoration = "manual"`. App history gives us an opportunity to provide some intermediate options to developers, at least for the case of same-document transitions. We do this via another option to `transitionWhile()`:
+
+- `e.transitionWhile(promise, { scrollRestoration: "after-transition" })`: the default behavior. The browser delays its scroll restoration logic until `promise` fulfills; it will perform no scroll restoration if the promise rejects. If the user has scrolled during the transition then no scroll restoration will be performed.
```

Comment:
Do we really want to cancel scroll behavior if the user scrolls during the transition? I feel like the current behavior in a multi-page application is to disregard user scrolls while the page is navigating?

Reply:
I'd tend to agree: user scroll should be ignored. But this rule could apply to manual calls to `window.scrollTo` and similar APIs.

Reply:
The Chromium engineers pointed me to code which abandons scroll restoration (for MPA navs) when the user scrolls. I will do some black-box testing to confirm.

Reply:
Here is an example:

1. Open https://neat-equal-cent.glitch.me/, preferably in a vertically-small-ish window
2. Wait for, e.g., paragraph 50 to load, and scroll down to it
3. Click the example.com link
4. Click back. Do not touch anything.
5. After paragraph 50 loads, scroll position will be restored.

Then try it this way:

4. Click back. Once a scrollbar appears, scroll slightly.
5. Even after paragraph 50 loads, your scroll position will not be disturbed. You having scrolled will have changed how it works.